api/valor_api/backend/core/llm_clients.py

```diff
@@ -0,0 +1,641 @@
+        list[dict[str,str]]
+            The list of verdicts for each statement. Each verdict is a dictionary with the "verdict" and optionally a "reason".
+        """
+        messages = [
+            {"role": "system", "content": DEFAULT_SYSTEM_PROMPT},
+            {
+                "role": "user",
+                "content": _generate_anwswer_relevance_verdicts_instruction(
+                    query, statements
+                ),
+            },
+        ]
+
+        response = self(messages)
+        response = trim_and_load_json(response)
+        if type(response) != dict or "verdicts" not in response:
+            raise InvalidLLMResponseError(
+                f"LLM response was not a valid list of verdicts: {response}"
+            )
+
+        verdicts = response["verdicts"]
+        if (
+            type(verdicts) != list
+            or len(verdicts) != len(statements)
+            or not all(
+                verdict["verdict"] in ["yes", "no", "idk"]
+                for verdict in verdicts
+            )
+        ):
+            raise InvalidLLMResponseError(
+                f"LLM response was not a valid list of verdicts: {response}"
+            )
+
+        return verdicts
+
+    def answer_relevance(
+        self,
+        query: str,
+        text: str,
+    ) -> float:
+        """
+        Compute answer relevance, the proportion of statements that are relevant to the query, for a single piece of text.
+
+        Parameters
+        ----------
+        query: str
+            The query to evaluate the statements against.
+        text: str
+            The text to extract statements from.
+
+        Returns
+        -------
+        float
+            The answer relevance score will be evaluated as a float between 0 and 1, with 1 indicating that all statements are relevant to the query.
+        """
+        statements = self._generate_statements(text)
+        verdicts = self._generate_anwswer_relevance_verdicts(query, statements)
+        return sum(
+            1 for verdict in verdicts if verdict["verdict"] == "yes"
+        ) / len(verdicts)
+
+    def coherence(
+        self,
+        text: str,
+    ) -> int:
+        """
+        Compute coherence, the collective quality of all sentences, for a single piece of text.
+
+        Parameters
+        ----------
+        text: str
+            The text to be evaluated.
+
+        Returns
+        -------
+        int
+            The coherence score will be evaluated as an integer, with 1 indicating the lowest coherence and 5 the highest coherence.
+        """
+        messages = [
+            {"role": "system", "content": DEFAULT_SYSTEM_PROMPT},
+            {"role": "user", "content": _get_coherence_instruction(text)},
+        ]
+
+        response = self(messages)
+
+        try:
+            # Valid responses: "5", "\n5", "5\n", "5.", " 5", "5 {explanation}", etc.
+            ret = int(response.strip()[0])
+            assert ret in {1, 2, 3, 4, 5}
```

Comment:
`assert` should not be used in production code as it can be ignored depending on python interpreter flags. See [link](https://stackoverflow.com/questions/944592/best-practice-for-using-assert).

This should be changed into explicit exception, it may be good to create a custom exception in `valor_api.exceptions`. A possible exception could be `MalformedLLMResponseError` which you could pass the response into. This would inform the user what the model responded with.